dvc/repo/diff.py

```diff
@@ -0,0 +1,189 @@
+    result = {
+        DIFF_DEL: 0,
+        DIFF_IDENT: 0,
+        DIFF_CHANGE: 0,
+        DIFF_NEW: 0,
+        DIFF_MOVE: 0,
+    }
+    keys = set(a_entries.keys())
+    keys.update(b_entries.keys())
+    diff_size = 0
+    for key in keys:
+        if key in a_entries and key in b_entries:
+            if a_entries[key] == b_entries[key]:
+                result[DIFF_IDENT] += 1
+            else:
+                result[DIFF_CHANGE] += 1
+                diff_size += os.path.getsize(
+                    self.cache.local.get(b_entries[key])
+                ) - os.path.getsize(self.cache.local.get(a_entries[key]))
+        elif key in a_entries:
+            result[DIFF_DEL] += 1
+            diff_size -= os.path.getsize(self.cache.local.get(a_entries[key]))
+        else:
+            result[DIFF_NEW] += 1
+            diff_size += os.path.getsize(self.cache.local.get(b_entries[key]))
+    result[DIFF_SIZE] = diff_size
+    return result
+
+
+def _get_diff_outs(self, diff_dct):
+    self.tree = diff_dct[DIFF_A_TREE]
+    a_outs = {str(out): out for st in self.stages() for out in st.outs}
+    self.tree = diff_dct[DIFF_B_TREE]
+    b_outs = {str(out): out for st in self.stages() for out in st.outs}
+    outs_paths = set(a_outs.keys())
+    outs_paths.update(b_outs.keys())
+    results = {}
+    for path in outs_paths:
+        if path in a_outs and path in b_outs:
+            results[path] = {
+                DIFF_A_OUTPUT: a_outs[path],
+                DIFF_B_OUTPUT: b_outs[path],
+                DIFF_NEW_FILE: False,
+                DIFF_DELETED: False,
+                # possible drawback: regular file ->directory movement
+                DIFF_IS_DIR: a_outs[path].is_dir_cache,
+            }
+        elif path in a_outs:
+            results[path] = {
+                DIFF_A_OUTPUT: a_outs[path],
+                DIFF_NEW_FILE: False,
+                DIFF_DELETED: True,
+                DIFF_IS_DIR: a_outs[path].is_dir_cache,
+            }
+        else:
+            results[path] = {
+                DIFF_B_OUTPUT: b_outs[path],
+                DIFF_NEW_FILE: True,
+                DIFF_DELETED: False,
+                DIFF_IS_DIR: b_outs[path].is_dir_cache,
+            }
+    return results
+
+
+def _diff_dir(self, target, diff_dct):
+    result = {DIFF_TARGET: target}
+    result[DIFF_IS_DIR] = True
+    a_entries, b_entries = {}, {}
+    try:
+        if not diff_dct[DIFF_NEW_FILE]:
+            result[DIFF_OLD_FILE] = target
+            result[DIFF_OLD_CHECKSUM] = diff_dct[DIFF_A_OUTPUT].checksum
+            a_entries = _extract_dir(self, diff_dct[DIFF_A_OUTPUT])
+        if not diff_dct[DIFF_DELETED]:
+            result[DIFF_NEW_FILE] = target
+            result[DIFF_NEW_CHECKSUM] = diff_dct[DIFF_B_OUTPUT].checksum
+            b_entries = _extract_dir(self, diff_dct[DIFF_B_OUTPUT])
+        result.update(_get_tree_changes(self, a_entries, b_entries))
+    except FileNotFoundError:
```

Comment:
In some older python2.7 versions there is no FileNotFoundError. So for compatibility, we need to use OSError and then check that errno is ENOENT as an equivalent to FileNotFoundError.